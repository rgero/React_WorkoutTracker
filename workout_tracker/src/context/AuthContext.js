import AsyncStorage from '@react-native-async-storage/async-storage';

import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const authReducer = (state, action) => {
    switch (action.type)
    {
        case "addError":
            return { ...state, errorMessage: action.payload};
        case "storeToken":
            return { errorMessage: "", token: action.payload};
        case "clearErrorMessage":
            return {...state, errorMessage: ""}
        case "signout":
            return { token: null, errorMessage: ""}
        default:
            return state;
    }
};

const tryLocalSignin = dispatch => async () => {
    const token = await AsyncStorage.getItem('token');
    if (token)
    {
        dispatch({ type: 'storeToken', payload: token})
    }
}

const clearErrorMessage = dispatch => () => {
    dispatch( {type: 'clearErrorMessage' })
}

const signUp = dispatch => {
    return async ({ email, displayName, password }) => {
        try {
            const response = await trackerAPI.post('/signup', { email, displayName, password });
            
            // Storing the Token
            await AsyncStorage.setItem('token', response.data.token);
            dispatch({type: "storeToken", payload: response.data.token})
        } catch (err) {
            console.log(err.message);
            dispatch({ type: 'addError', payload: "Sign-up failed. Please try again"})
        }
    };
}

const signIn = (dispatch) => {
    return async ({ email, password }) => {
        try {
            const response = await trackerAPI.post('/signin', { email, password });
            
            // Storing the Token
            await AsyncStorage.setItem('token', response.data.token);
            dispatch({type: "storeToken", payload: response.data.token})
        } catch (err) {
            console.log(err.message);
            dispatch({ type: 'addError', payload: "Login failed. Please try again"})
        }
    };
}

const signOut = dispatch => async () => {
    try {
        await AsyncStorage.removeItem('token');
        dispatch({type: "signout"})
    } catch (err)
    {
        console.log(err);
    }
}

export const {Provider, Context } = createDataContext(authReducer, 
                                                      {clearErrorMessage, signUp, signIn, signOut, tryLocalSignin}, 
                                                      {
                                                        token: null,
                                                        errorMessage: ""
                                                      }
                                                    )