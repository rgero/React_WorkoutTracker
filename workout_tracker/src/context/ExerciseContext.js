import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const exerciseReducer = (state, action) => {
    switch(action.type)
    {
        case 'fetch':
            return action.payload;
        default:
            return state;
    }
};

const fetchExercises = dispatch => async () => {
    const response = await trackerAPI.get('/exercises');
    dispatch({type: 'fetch', payload: response.data})
};

const fetchExerciseByID = dispatch => async (targetID) => {
    const response = await trackerAPI.get(`/exercises/${targetID}`);
    dispatch({type: 'fetch', payload: response.data})
};


export const {Provider, Context} = createDataContext(
    exerciseReducer,
    {fetchExercises, fetchExerciseByID},
    []
);