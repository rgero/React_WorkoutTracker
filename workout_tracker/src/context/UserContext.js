import AsyncStorage from '@react-native-async-storage/async-storage';

import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const userReducer = (state, action) => {
    switch(action.type)
    {
        case 'fetch':
            return action.payload;
        default:
            return state;
    }
};

const getUser = dispatch => async () => {
    const token = await AsyncStorage.getItem('token');
    if (token)
    {
        const response = await trackerAPI.get('/users');
        dispatch({type: 'fetch', payload: response.data})
    }
}

export const {Provider, Context} = createDataContext(
    userReducer,
    {getUser},
    []
);