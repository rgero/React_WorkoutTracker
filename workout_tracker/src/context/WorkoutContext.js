import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const workoutReducer = (state, action) => {
    switch(action.type)
    {
        case 'fetch':
            return action.payload;
        case "create":
            return state.concat(action.payload);
        case "edit":
            return state.map( (workout) => {
                if (workout._id === action._id) {
                    return {
                        ...workout,
                        ...action.update
                    }
                } else {
                    return workout;
                }
            })
        case "delete":
            return state.filter( (testWorkout) => {
                return action.id !== testWorkout._id;
            });
        default:
            return state;
    }
};

const fetchWorkouts = dispatch => async () => {
    const response = await trackerAPI.get('/workouts');
    dispatch({type: 'fetch', payload: response.data})
};

const createWorkout = dispatch => async (workout) => 
{
    const response = await trackerAPI.post('/workouts', workout)
    dispatch({type: "create", payload: response.data});
};

const updateWorkout = dispatch => async (workout) =>
{
    const response = await trackerAPI.put(`/workouts/${workout._id}`, workout);
    dispatch({type: "edit", payload: response.data});
}

const deleteWorkout = dispatch => async (targetID) => 
{
    const response = await trackerAPI.delete(`/workouts/${targetID}`)
    dispatch({type: "delete", id: targetID});
}

export const {Provider, Context} = createDataContext(
    workoutReducer,
    {fetchWorkouts, createWorkout, deleteWorkout, updateWorkout},
    []
);