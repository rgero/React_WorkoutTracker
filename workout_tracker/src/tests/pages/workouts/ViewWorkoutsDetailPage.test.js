import React from 'react';
import TestRenderer from 'react-test-renderer';
import reactRouterDom from 'react-router-dom';
import { act } from 'react-dom/test-utils';


import {Context as WorkoutContext} from '../../../context/WorkoutContext';
import ViewWorkoutDetailsPage from '../../../pages/workouts/ViewWorkoutDetailPage';
import {TestWorkouts} from '../../testData';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: jest.fn(),
    useNavigate: jest.fn()
}));


test('Rendering a Workout', async ()=> {

    const testContextValues = {
        state: TestWorkouts,
        fetchWorkouts: jest.fn(),
        deleteWorkout: jest.fn()
    }

    jest.spyOn(reactRouterDom, 'useParams').mockReturnValue({ id: '1234' });

    let element;
    await act(async () => {
        element = new TestRenderer.create(
            <WorkoutContext.Provider value={testContextValues}>
                <ViewWorkoutDetailsPage/>
            </WorkoutContext.Provider>
        );
    })
    
    expect(element.toJSON()).toMatchSnapshot();
})

test('Rendering a failure', async ()=> {

    const testContextValues = {
        state: TestWorkouts,
        fetchWorkouts: jest.fn(),
        deleteWorkout: jest.fn()
    }

    jest.spyOn(reactRouterDom, 'useParams').mockReturnValue({ id: '123123123' });

    let element;
    await act(async () => {
        element = new TestRenderer.create(
            <WorkoutContext.Provider value={testContextValues}>
                <ViewWorkoutDetailsPage/>
            </WorkoutContext.Provider>
        );
    })
    
    expect(element.toJSON()).toMatchSnapshot();
})