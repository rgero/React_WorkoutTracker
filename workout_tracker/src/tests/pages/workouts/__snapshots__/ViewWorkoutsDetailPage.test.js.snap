// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering a Workout 1`] = `
<div
  className="pt-4 container-md"
>
  <div
    className="align-items-center row"
  >
    <div
      className="col"
    >
      <h1>
        Workout on 
        2023-03-07
      </h1>
    </div>
    <div
      className="col"
    >
      <div
        className="float-end dropdown"
      >
        <button
          aria-expanded={false}
          className="dropdown-toggle btn btn-primary"
          disabled={false}
          id="dropdown-basic"
          onClick={[Function]}
          type="button"
        >
          Options
        </button>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="card"
    >
      <div
        className="card-body"
      >
        <div
          className="card-title h5"
        >
          Notes
        </div>
        <div
          className="card-body"
        >
          Testing Notes for Testing
        </div>
      </div>
    </div>
    <div
      className="card"
    >
      <div
        className="card-body"
      >
        <div
          className="card-title h5"
        >
          Exercises - 
          2
        </div>
        <div
          className="card-body"
        >
          <div
            className="container"
          >
            <div
              className="list-group list-group-flush"
              onKeyDown={[Function]}
              style={
                Object {
                  "margin": "auto",
                  "width": "60%",
                }
              }
            >
              <div
                className="list-group-item"
                data-rr-ui-event-key={null}
                onClick={[Function]}
                style={
                  Object {
                    "border": "0px",
                  }
                }
              >
                <div
                  className="card"
                >
                  <div
                    className="card-body"
                  >
                    <div
                      className="card-title h5"
                    >
                      #
                      1
                       - 
                      Bench Press
                    </div>
                    <div
                      className="card-body"
                    >
                      <div
                        className="pt-3 container"
                        style={
                          Object {
                            "margin": "auto",
                            "textAlign": "center",
                            "width": "60%",
                          }
                        }
                      >
                        <div
                          style={
                            Object {
                              "textAlign": "left",
                            }
                          }
                        >
                          Set List
                        </div>
                        <table
                          className="table table-sm table-striped table-bordered table-hover"
                        >
                          <thead>
                            <tr>
                              <th>
                                Reps
                              </th>
                              <th>
                                Weight
                              </th>
                            </tr>
                          </thead>
                          <tbody>
                            <tr
                              onClick={[Function]}
                            >
                              <td>
                                20
                              </td>
                              <td>
                                275
                              </td>
                            </tr>
                            <tr
                              onClick={[Function]}
                            >
                              <td>
                                20
                              </td>
                              <td>
                                300
                              </td>
                            </tr>
                            <tr
                              onClick={[Function]}
                            >
                              <td>
                                3
                              </td>
                              <td>
                                315
                              </td>
                            </tr>
                          </tbody>
                        </table>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="list-group-item"
                data-rr-ui-event-key={null}
                onClick={[Function]}
                style={
                  Object {
                    "border": "0px",
                  }
                }
              >
                <div
                  className="card"
                >
                  <div
                    className="card-body"
                  >
                    <div
                      className="card-title h5"
                    >
                      #
                      2
                       - 
                      Shrugs
                    </div>
                    <div
                      className="card-body"
                    >
                      <div
                        className="pt-3 container"
                        style={
                          Object {
                            "margin": "auto",
                            "textAlign": "center",
                            "width": "60%",
                          }
                        }
                      >
                        <div
                          style={
                            Object {
                              "textAlign": "left",
                            }
                          }
                        >
                          Set List
                        </div>
                        <table
                          className="table table-sm table-striped table-bordered table-hover"
                        >
                          <thead>
                            <tr>
                              <th>
                                Reps
                              </th>
                              <th>
                                Weight
                              </th>
                            </tr>
                          </thead>
                          <tbody>
                            <tr
                              onClick={[Function]}
                            >
                              <td>
                                40
                              </td>
                              <td>
                                100
                              </td>
                            </tr>
                            <tr
                              onClick={[Function]}
                            >
                              <td>
                                40
                              </td>
                              <td>
                                100
                              </td>
                            </tr>
                            <tr
                              onClick={[Function]}
                            >
                              <td>
                                40
                              </td>
                              <td>
                                100
                              </td>
                            </tr>
                            <tr
                              onClick={[Function]}
                            >
                              <td>
                                40
                              </td>
                              <td>
                                100
                              </td>
                            </tr>
                          </tbody>
                        </table>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="card-body"
        />
      </div>
    </div>
    <div
      className="pt-2 col"
    >
      <a
        className="float-end btn btn-primary"
        disabled={false}
        href="/dashboard"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        Return to Dashboard
      </a>
    </div>
  </div>
</div>
`;

exports[`Rendering a failure 1`] = `
<div
  className="container"
>
  <div
    className="fade alert alert-danger show"
    role="alert"
  >
    <div
      className="alert-heading h4"
    >
      Oh snap! You got an error!
    </div>
    <p>
      An error has occurred
    </p>
    <div
      className="d-flex justify-content-end"
    >
      <button
        className="btn btn-outline-danger"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Return to Dashboard
      </button>
    </div>
  </div>
</div>
`;
